// app/routes/problems.js
import express from 'express';
import Problem from "app/models/Problem";
import { randomIntRange, shuffle } from 'okrand';
const router = express.Router();

const MAX_NUMBER = 9;


function create(req, res) {

    console.log('Creating Problem: ',req.body);
    let problem = new Problem(req.body);
    problem.save(function (err) {
        if (err) {
            console.log('error',err);
            res.send({error: 'something bad happened'});
        } else {

            res.send({problem: problem});
        }
    });
}

function answer(req, res) {
    let body = req.body;
    let id = body.id;
    let answer = body.answer;

    res.send({isCorrect});
}

// for later use
/*
function answerDb(req, res) {
    let body = req.body;
    let id = body.id;
    let answer = body.answer;
    Problem.findById(id).then(problem=>{
        let isCorrect = (problem.correctAnswer ===  answer);
        res.send({isCorrect});
    });

}
*/

function random(req, res) {
    const a = randomIntRange(0, MAX_NUMBER);
    const b = randomIntRange(0, MAX_NUMBER);
    const correctAnswer = a + b;

    const problem = new Problem({
        question: `${a}+${b}`,
        correctAnswer: correctAnswer,
        wrongAnswers: create3WrongAnswers(correctAnswer),

    });
    const options = problem.buildOptions();
    res.send({problem, options});
}

const createWrongAnswer = (correctAnswer, except) => {
    let badTest = true;
    let test = null;
    while(badTest) {
        test = randomIntRange(0,MAX_NUMBER*2);
        badTest = ( (test === correctAnswer) || except.includes(test) );
    }
    return test;
};

const create3WrongAnswers = (correctAnswer) => {
    const first = createWrongAnswer(correctAnswer, []);
    const second = createWrongAnswer(correctAnswer, [first]);
    const third = createWrongAnswer(correctAnswer, [first,second]);
    return [first, second, third];
};


// For later use
function randomDb(req, res) {
    Problem.find({}, function(err, problems){
        if(err){
            console.log('error getting problems: ',err);
            res.send({error: 'something bad happened'});
        } else {
            //console.log(problems);
            let length = problems.length;
            if(length === 0) {
                throw new Error('No Problems');
            }
            let randomIndex = randomIntRange(0,length-1);
            console.log('randomIndex: ',randomIndex);
            let problem  = problems[randomIndex];
            let options = problem.buildOptions();
            res.send({problem, options});
        }
    })
}

function list(req, res) {
    Problem.find({}, function(err, problems){
        if(err){
            console.log('error getting problems: ',err);
            res.send({error: 'something bad happened'});
        } else {
            res.send(problems);
        }
    });

}

router.post('/', create);

router.get('/', list);

router.get('/random', random);
router.post('/answer', answer);


export default router;
